client.go

```diff
@@ -0,0 +1,593 @@
+		return &Action{Code: ActContinue}, nil
+	}
+
+	msg := &Message{
+		Code: byte(CodeHeader),
+	}
+	msg.Data = appendCString(msg.Data, key)
+	msg.Data = appendCString(msg.Data, value)
+
+	if err := writePacket(s.conn, msg, s.writeTimeout); err != nil {
+		return nil, fmt.Errorf("milter: header field: %w", err)
+	}
+
+	act, err := s.readAction()
+	if err != nil {
+		return nil, fmt.Errorf("milter: header field: %w", err)
+	}
+	return act, nil
+}
+
+// HeaderEnd send the EOH (End-Of-Header) message to the milter.
+//
+// No HeaderField calls are allowed after this point.
+func (s *ClientSession) HeaderEnd() (*Action, error) {
+	if s.protoMask&OptNoEOH != 0 {
+		return &Action{Code: ActContinue}, nil
+	}
+
+	if err := writePacket(s.conn, &Message{
+		Code: byte(CodeEOH),
+	}, s.writeTimeout); err != nil {
+		return nil, fmt.Errorf("milter: header end: %w", err)
+	}
+
+	act, err := s.readAction()
+	if err != nil {
+		return nil, fmt.Errorf("milter: header end: %w", err)
+	}
+	return act, nil
+}
+
+// Header sends each field from textproto.Header followed by EOH unless
+// header messages are disabled during negotiation.
+func (s *ClientSession) Header(hdr textproto.Header) (*Action, error) {
+	for f := hdr.Fields(); f.Next(); {
+		act, err := s.HeaderField(f.Key(), f.Value())
+		if err != nil {
+			return nil, err
+		}
+
+		if act.Code != ActContinue {
+			return act, nil
+		}
+	}
+
+	return s.HeaderEnd()
+}
+
+// BodyChunk sends a single body chunk to the milter.
+//
+// It is callers responsibility to ensure every chunk is not bigger than
+// MaxBodyChunk.
+func (s *ClientSession) BodyChunk(chunk []byte) (*Action, error) {
+	if s.protoMask&OptNoBody != 0 {
+		return &Action{Code: ActContinue}, nil
+	}
+
+	// Callers tend to be irresponsible... /s
+	if len(chunk) > MaxBodyChunk {
+		return nil, fmt.Errorf("milter: body chunk: too big body chunk: %v", len(chunk))
+	}
+
+	if err := writePacket(s.conn, &Message{
+		Code: byte(CodeBody),
+		Data: chunk,
+	}, s.writeTimeout); err != nil {
+		return nil, fmt.Errorf("milter: body chunk: %w", err)
+	}
+
+	act, err := s.readAction()
+	if err != nil {
+		return nil, fmt.Errorf("milter: body chunk: %w", err)
+	}
+	return act, nil
+}
+
+// Body is a helper function that calls BodyChunk repeately to transmit entire
+// body from io.Reader and then calls End.
+//
+// See documentation for these functions for details.
+func (s *ClientSession) Body(r io.Reader) ([]ModifyAction, *Action, error) {
+	// It is problematic to use io.WriteCloser since we may need to report
+	// action after each write.
+
+	buf := make([]byte, MaxBodyChunk)
+	for {
+		n, err := r.Read(buf)
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return nil, nil, err
+		}
+		if n == 0 {
+			break
+		}
+
+		act, err := s.BodyChunk(buf[:n])
+		if err != nil {
+			return nil, nil, err
+		}
+		if act.Code != ActContinue {
+			return nil, act, nil
+		}
+	}
+
+	return s.End()
+}
+
+type ModifyAction struct {
```

Comment:
I wonder if one struct per action would be better than a single one-size-fits-all struct.

Reply:
I don't have any strong opinions but one structure seems somewhat easier to work with.